---
layout: default
image: /img/bg.jpg
sectionid: home
desc: Welcome to the JEFworks Lab,<br> where Prof. Jean Fan and team work on computational software and statistical approaches to address questions in developmental and cancer biology. <br><br> <a href="join"><button type="button" class="btn btn-warning btn-lg">Recruiting at all levels! Join us!</button></a>
---

<div class="container">
    <div class="row">

	<h2>We are a bioinformatics research lab in the <a href="https://www.bme.jhu.edu/">Department of Biomedical Engineering</a> at Johns Hopkins University. We are also a part of the <a href="https://ccb.jhu.edu/">Center for Computational Biology</a> and <a href="https://www.cs.jhu.edu/">Department of Computer Science</a>.</h2>
	<div class="row" style="padding: 15px">
	    <!--- <a href="team"><img src="img/team.jpg" width="100%"></a> --->
	    <div class="col-xs-4 col-md-2">
		<a href="team"><img src="/img/profile.jpg" class="img-responsive img-circle" alt="Jean Fan" style="padding: 15px 5px"></a>
	    </div>
	    <div id="my-team">
		{% for pub in site.team %}
		<div class="col-xs-4 col-md-2">
		    <a href="team">
			<img src="{{ pub.profile }}" class="img-responsive img-circle" alt="{{ pub.name }}" style="padding: 15px 5px">
		    </a>
		</div>
		{% endfor %}
	    </div>
	    
	</div>
	<h2>We develop methods for analyzing single-cell spatially resolved transcriptomic sequencing and imaging data.</h2>
	<div class="row">
	    <div class="col-md-7">
		<p>Spatial organization at both the subcellular-level within cells as well as the cellular-level within tissues play important roles in regulating cell identity and function. Recent technological advances have enabled high-throughput spatially resolved transcriptomic profiling at single-molecule and near-single-cell resolution. We develop machine learning and other statistical approaches as open-source computational software to take advantage of this new spatial information in deriving biological insights regarding how spatial organization plays a role in both healthy and diseased settings. </p>
		<ul>
		    <li>
			<a href="/publications#/papers/2021/09/06/Atta_et_al-2021-Nature_Communications/"><u>Lyla Atta</u>, <u>Jean Fan^</u>. Computational challenges and opportunities in spatially resolved transcriptomic data analysis. Nature Communications. September 6, 2021. doi.org/10.1038/s41467-021-25557-9</a>
		    <li>
			<a href="/publications#/papers/2021/06/16/biorxiv448381/"><u>Brendan F Miller</u>, <u>Lyla Atta</u>, <u>Arpan Sahoo</u>, <u>Feiyang Huang</u>, <u>Jean Fan^</u>. Reference-free cell-type deconvolution of pixel-resolution spatially resolved transcriptomics data. bioRxiv. 2021. doi: 10.1101/2021.06.15.448381</a>
		    </li>
		    <li>
			<a href="/publications#/papers/2021/09/28/Atta_et_al-2021-Bioinformatics/"><u>Lyla Atta</u>, <u>Arpan Sahoo</u>, <u>Jean Fan^</u>. VeloViz: RNA-velocity informed embeddings for visualizing cellular trajectories. Bioinformatics. 2021. https://doi.org/10.1093/bioinformatics/btab653</a>
		    </li>
		    <li>
			<a href="/publications#/papers/2021/05/25/gr271288120/"><u>Brendan F Miller</u>, Dhananjay Bambah-Mukku, Catherine Dulac, Xiaowei Zhuang, <u>Jean Fan^</u>. Characterizing spatial gene expression heterogeneity in spatially resolved single-cell transcriptomics data with nonuniform cellular densities. Genome Research. 2021. doi:10.1101/gr.271288.120</a>
		    </li>
		    <li>
			<a href="/publications#/papers/2019/09/09/pnas1912459116/">Chenglong Xia*, <u>Jean Fan*</u>, George Emanuel*, Junjie Hao, and Xiaowei Zhuang. Spatial transcriptome profiling by MERFISH reveals subcellular RNA compartmentalization and cell cycle-dependent gene expression. PNAS. 2019. doi:10.1073/pnas.1912459116</a>
		    </li>
		</ul>
	    </div>
	    <div class="col-md-5">
		<img src="assets/papers/Atta_et_al-2021-Nature_Communications.png" width="100%">
	    </div>
	</div>

	<div class="row">
	    <h2>We apply these methods to better understand the impact of cellular heterogeneity on cancer pathogenesis and prognosis.</h2>
	    <div class="col-md-6">
		<img src="img/main1.jpg" width="100%">
	    </div>
	    <div class="col-md-6">
		<p>Advancements in high-throughput sequencing and imaging technologies have uncovered tremendous genetic, epigenetic, transcriptional, and spatial heterogeneity in various cancers but their impact on clinical outcomes is not well understood. We establish close collaborations with clinical collaborators to develop and apply bioinformatics methods that contribute to a more complete understanding of how cellular heterogeneity impacts tumor progression, therapeutic resistance, and ultimately clinical prognosis. We are particularly interested in pediatric gliomas.</p>
		<ul>
		    <li>
			<a href="/publications#/papers/2020/09/15/natureemm101038/"><u>Jean Fan^</u>, Kamil Slowikowski, Fan Zhang. Single-cell transcriptomics in cancer - computational challenges and opportunities. Nature Experimental and Molecular Medicine. Sept 15, 2020, doi.org:10.1038/s12276-020-0422-0</a>
		    </li>
		    <li>
			<a href="/publications#/papers/2018/06/13/gr228080117/"><u>Jean Fan*</u>, Hae-Ock Lee*, Soohyun Lee, Da-eun Ryu, Semin Lee, et al. Linking transcriptional and genetic tumor heterogeneity through allele analysis of single-cell RNA-seq. Genome Research. 2018. doi:10.1101/gr.228080.117</a>
		    </li>
		    <li>
			<a href="/publications#/papers/2017/05/22/gr217331116/">Lili Wang*, <u>Jean Fan*</u>, Joshua M. Francis, George Georghiou, Sarah Hergert, et al. Integrated single-cell genetic and transcriptional analysis suggests novel drivers of chronic lymphocytic leukemia. Genome Research. 2017. doi:10.1101/gr.217331.116</a>
		    </li>
		</ul>
	    </div>
	</div>
    </div>
</div>

<hr>

<div class="container">
    <div class="row">
	<div class="col-sm-4">
	    <h1 class="text-center"><i class="fa fa-pencil" aria-hidden="true"></i></h1>
	    <h2 class="text-center">Latest Publications</h2>
	    <ul class="post-list-container">
		{% assign items = site.papers | reverse %}
		{% for post in items limit:5 %}
		<li><a href="{{ site.baseurl }}/publications#{{ post.url }}">{{ post.title }}</a> on {{ post.date | date_to_long_string }}</li>
		{% endfor %}
	    </ul>
	    <a href="{{ site.baseurl }}/publications">
		<button type="button" class="btn btn-primary btn-block">
		    More Publications
		</button>
	    </a>
	</div>
	<div class="col-sm-4">
	    <h1 class="text-center"><i class="fa fa-rss-square" aria-hidden="true"></i></h1>
	    <h2 class="text-center">Latest News</h2>
	    <ul class="post-list-container">
		{% assign items = site.news | reverse %}
		{% for post in items limit:5 %}
		<li><a href="{{ post.url | prepend: site.baseurl }}">{{ post.title }}</a> on {{ post.date | date_to_long_string}}</li>
		{% endfor %}
	    </ul>	    
	    <a href="{{ site.baseurl }}/allnews">
		<button type="button" class="btn btn-primary btn-block">
		    More News
		</button>
	    </a>
	</div>
	<div class="col-sm-4">
	    <h1 class="text-center"><i class="fa fa-cogs" aria-hidden="true"></i></h1>
	    <h2 class="text-center">Latest Blog Posts</h2>
	    <ul class="post-list-container">
		{% for post in site.posts limit:10 %}
		<li><a href="{{ post.url | prepend: site.baseurl }}">{{ post.title }}</a> on {{ post.date | date_to_long_string }}</li>
		{% endfor %}
	    </ul>
	    <a href="{{ site.baseurl }}/allposts"> 
		<button type="button" class="btn btn-primary btn-block">
		    More Blog Posts
		</button>
	    </a>
	</div>
    </div>

</div>

<hr>

<div class="container">
    <div class="row">

	<h2>Latest Lab Fun</h2>
	{% include carousel.html height="45" unit="%" duration="10" %}

    </div>
</div>

