---
layout: paper
title: Evidence of off-target probe binding in the 10x Genomics Xenium v1 Human Breast Gene Expression Panel compromises accuracy of spatial transcriptomic profiling
authors: <b><u>Caleb Hallinan</b></u>, Hyun Joo Ji, Steven L Salzberg, <b>Jean Fan^</b>
image: /assets/papers/opt.jpg
abstract: The accuracy of spatial gene expression profiles generated by probe-based in situ spatially-resolved transcriptomic technologies depends on the specificity with which probes bind to their intended target gene. Off-target binding, defined as a probe binding to something other than the target gene, can distort a gene’s true expression profile, making probe specificity essential for reliable transcriptomics. Here, we investigate off-target binding in the 10x Genomics Xenium v1 Human Breast Gene Expression Panel. We developed a software tool, Off-target Probe Tracker (OPT), to identify putative off-target binding via alignment of probe sequences and found at least 21 out of the 280 genes in the panel impacted by off-target binding to protein-coding genes. To substantiate our predictions, we leveraged a previously published Xenium breast cancer dataset generated using this gene panel and compared results to orthogonal spatial and single-cell transcriptomic profiles from Visium CytAssist and 3ʹ single-cell RNA-seq derived from the same tumor block. Our findings indicate that for some genes, the expression patterns detected by Xenium demonstrably reflect the aggregate expression of the target and predicted off-target genes based on Visium and single-cell RNA-seq rather than the target gene alone. Overall, this work enhances the biological interpretability of spatial transcriptomics data and improves reproducibility in spatial transcriptomics research.
journal: bioRxiv
journalurl: https://www.biorxiv.org/content/10.1101/2025.03.31.646342v1.abstract
code: https://github.com/JEFworks-Lab/off-target-probe-tracker
---

